### PR TITLE
Add errorsource to errors

### PR DESCRIPTION
**Pull request by iwysiu**
**What this PR does / why we need it**:
Adds errorsource to the opensearch datasource.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/oss-plugin-partnerships/issues/1014

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/opensearch/opensearch.go [hunk 1]
```diff
@@ -60,25 +61,25 @@ func (ds *OpenSearchDatasource) QueryData(ctx context.Context, req *backend.Quer
...
 	return response, err
 }
 
 // handleServiceMapPrefetch inspects the given request, and, if it wants a serviceMap, creates and
 // calls the Prefetch query to get the services and operations lists that are required for
 // the associated Stats query. It then adds these parameters to the originating query so
-// the Stats query can be created later.
-func handleServiceMapPrefetch(ctx context.Context, osClient client.Client, req *backend.QueryDataRequest) (string, error) {
+// the Stats query can be created later. Returns true if the serviceMap was fetched
+func handleServiceMapPrefetch(ctx context.Context, osClient client.Client, req *backend.QueryDataRequest) *backend.QueryDataResponse {
```
Comment (iwysiu) [inline pkg/opensearch/opensearch.go:78]: There wasn't really a better way to handle the fact that `execute` is now returning the error as part of the response and still wrap the other errors

File: pkg/opensearch/opensearch.go [hunk 2]
```diff
@@ -99,41 +102,33 @@ func handleServiceMapPrefetch(ctx context.Context, osClient client.Client, req *
...
+		response := backend.NewQueryDataResponse()
+		err = errorsource.SourceError(source, fmt.Errorf(`Error fetching service map info: %w`, err), true)
+		return errorsource.AddErrorToResponse(refId, response, err)
 	}
-	return nil, err
+	return nil
 }
 
 func wrapError(response *backend.QueryDataResponse, err error) (*backend.QueryDataResponse, error) {
-	var invalidQueryTypeError invalidQueryTypeError
```
Comment (iwysiu) [inline pkg/opensearch/opensearch.go:124]: now the error gets wrapped earlier, so we don't need this

File: pkg/opensearch/lucene_handler.go [hunk 3]
```diff
@@ -279,14 +280,16 @@ func (h *luceneHandler) executeQueries(ctx context.Context) (*backend.QueryDataR
...
 	if err != nil {
-		return nil, err
+		return errorsource.AddPluginErrorToResponse(h.queries[0].RefID, response, err), nil
 	}
 
 	res, err := h.client.ExecuteMultisearch(ctx, req)
 	if err != nil {
-		return nil, err
+		// We are returning the error containing the source that was added through errorsource.Middleware
+		return errorsource.AddErrorToResponse(h.queries[0].RefID, response, err), nil
```
Comment (nmarrs) [inline pkg/opensearch/lucene_handler.go:292]: Should we maybe draw out the `h.queries[0].RefID` into a defined reusable var in this case?

File: pkg/opensearch/opensearch.go [hunk 4]
```diff
@@ -60,25 +61,25 @@ func (ds *OpenSearchDatasource) QueryData(ctx context.Context, req *backend.Quer
...
-	if err != nil {
-		return wrapServiceMapPrefetchError(errRefID, err)
+	response := handleServiceMapPrefetch(ctx, osClient, req)
+	if response != nil {
+		return response, nil
 	}
 
 	query := newQueryRequest(osClient, req.Queries, req.PluginContext.DataSourceInstanceSettings)
-	response, err := wrapError(query.execute(ctx))
+	response, err = wrapError(query.execute(ctx))
```
Comment (nmarrs) [inline pkg/opensearch/opensearch.go:70]: what is context of removing the `:=` to `=`? (I am a bit of a Go noob btw)

Comment (nmarrs) [reply pkg/opensearch/opensearch.go:70]: also should we keep this version of `wrapError` around vs using the newer wrap for this error? or in this case are we confident that this is an error owned by us

**Review by nmarrs — COMMENTED** [review]
nice!! just left some nooby go questions haha

**Review by nmarrs — APPROVED** [review]
🚢

**Review by njvrzm — APPROVED** [review]
Looks solid! I like the lots of tests, I don't think we've been doing much of that with the errorsource work so far. One optional suggestion.

**Pull request merged**